README.md
# Unique Random Numbers Generator (TypeScript / JavaScript)

Generates unique random numbers between a minimum and maximum (inclusive). Negative numbers supported.

Example:

min = -5

max = 8

sample output: -3,-1,3,0,5,6,2,-4,-5,8,4,-2,1,7

## Install Instructions

Best way is to use `npm`:

```
npm i --save uniquerand
```

## Usage Instructions

### TypeScript

For TypeScript need `@types/node` to use `require()`:

```
npm i --save-dev @types/node
```

To use `UniqueRand.ts` in your TypeScript project here is an example. Modify to suit your needs:

```
const { UniqueRand } = require("uniquerand");

let min = -5;
let max = 8;
let arr = UniqueRand.getRandArr(min, max);
console.log('unique random numbers = ' + arr);
```

Note that the member function getRandArr is static so no need to instantiate the UniqueRand class before use.

If numbers with decimals are given it will round the `min` to the next lowest int it closest to and for the `max` to next highest int it is closest to. Eg:

```
let min = 0.2; 
let max = 9.3;
let arr = UniqueRand.getRandArr(min, max);
console.log('unique random numbers = ' + arr);
```

The min and max will be rounded and will effectively be as if you used:

```
min = 0
max = 10
```

sample output: 8,0,10,3,6,4,1,5,2,9,7

You may also pass an optional `count` argument that specifies the number of random numbers to return eg:

```
let min = 0.2; 
let max = 9.3;
let count = 5;
let arr = UniqueRand.getRandArr(min, max, count);
console.log('unique random numbers = ' + arr);
```
sample output: 1,9,3,5,10

### JavaScript

Make sure the `.js` files are up to date in the `dist` folder:
```
npm run build
```

To use `UniqueRand.js` in your JavaScript project here is an example. Modify to suit your needs:

```
let { UniqueRand } = require("uniquerand");

let min = -5;
let max = 8;
let arr = UniqueRand.getRandArr(min, max);
console.log('unique random numbers = ' + arr);
```
Note that the member function getRandArr is static so no need to instantiate the UniqueRand class before use.

If numbers with decimals are given it will round the `min` to the next lowest int it closest to and for the `max` to next highest int it is closest to. Eg:

```
let min = 0.2; 
let max = 9.3;
let arr = UniqueRand.getRandArr(min, max);
console.log('unique random numbers = ' + arr);
```

The min and max will be rounded and will effectively be as if you used:

```
min = 0
max = 10
```

sample output: 8,0,10,3,6,4,1,5,2,9,7

You may also pass an optional `count` argument that specifies the number of random numbers to return eg:

```
let min = 0.2; 
let max = 9.3;
let count = 5;
let arr = UniqueRand.getRandArr(min, max, count);
console.log('unique random numbers = ' + arr);
```
sample output: 1,9,3,5,10

## TypeScript Testing Instructions

Change directory to the module uniquerand then install dependencies:

```
npm install
```

Run Jest tests:

```
npm run test
```

Note: above is an alias to `jest src`.

<details>
  <summary>Click to show example test run.</summary>

```
$ npm run test

> uniquerand@1.0.6 test
> jest src

 PASS  src/uniqueRand.test.ts
  ✓ all (46 ms)

  console.log
    runSingleTest(1, 1) = [1] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(1, 2) = [1,2] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(1, 3) = [3,1,2] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(0, 10) = [9,4,0,10,5,6,1,3,2,7,8] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(0.2, 9.3) = [5,3,8,9,10,1,2,4,0,6,7] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(-0.2, 9.3) = [9,5,2,0,10,-1,7,8,3,4,6,1] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(-20.2, -9.3) = [-11,-18,-15,-19,-16,-12,-13,-20,-21,-14,-9,-10,-17] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(-5, 8) = [-2,3,5,-3,2,7,0,-1,4,1,8,-5,6,-4] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(1, 15) = [13,10,5,4,7,11,9,8,15,12,14,2,1,6,3] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(5, 95) = [9,41,34,14,67,93,17,69,76,5,15,88,74,50,8,57,83,72,75,19,45,28,26,25,22,20,35,23,49,40,92,33,84,53,80,61,77,31,82,16,64,48,12,65,32,10,56,43,6,51,36,37,39,70,44,94,73,90,59,24,89,47,66,30,91,60,13,29,18,71,95,58,81,78,55,11,27,68,52,86,54,87,46,38,85,7,79,63,62,21,42] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:73:13)

  console.log
    runSingleTest(0, 10, 5) = [9,7,3,10,2] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(0.2, 9.3, 5) = [6,3,0,7,4] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(-0.2, 9.3, 5) = [8,0,10,6,-1] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(-20.2, -9.3, 5) = [-10,-17,-15,-18,-20] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(-5, 8, 5) = [-5,0,2,7,5] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(1, 15, 6) = [5,7,8,4,15,14] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(5, 95, 8) = [23,50,77,37,74,75,30,84] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(1, 5, 6) = [2,3,4,5,1] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(-2, 5, 88) = [3,2,-2,0,4,-1,5,1] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(1, 5, -8) = [] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(-8, 5, -8) = [] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

  console.log
    runSingleTest(8, -1, -8) = [] ... Passed

      at runSingleTest (src/uniqueRand.test.ts:71:13)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.369 s, estimated 2 s
Ran all test suites matching /src/i.
```
</details>
 
# Feedback

Any constructive feedback / suggestions / new feature requests welcome!


CHANGELOG
v1.0.6 2023.10.14
    * Removed redundant code in getUniqueRand().
    * Fixed bug in isUnique.
    * Fixed bug in getRandArr() with count exceeding max - min + 1.
    * Fixed bug in isUnique() not testing correctly for uniqueness.
    * Added to runSingleTest() checks for negative count and count exceeding max - min + 1.
    * Added to runSingleTest() test cases for negative count and count exceeding max - min + 1.

v1.0.5 2021.9.2
- Updated package.json main field to point to "dist/UniqueRand.js".
- Updated README.md for npm instructions.

v1.0.4 2021.8.31
- Added to npm.

v1.0.3 2021.8.30
- updated package.json git URLs.
- Added tsconfig.tsbuildinfo to .gitignore.
- Added tsconfig.tsbuildinfo to .npmignore.
- Changed name in package.json.

v1.0.2 2021.8.30
- UniqueRand.ts now uses static member functions.
- Changed testing framework to Jest.
- Updated README.md.
- added .npmignore.
- Changed name in package.json.

v1.0.1 2021.8.24
- Added count argument. 

v1.0.0 2021.8.24
- Converted to TypeScript.


src/uniqueRand.test.ts

import 'jest';

const { UniqueRand } = require('./UniqueRand');

function isUnique(arr: number[], min: number, max: number): boolean {

  let ok = true;

  min = Math.floor(min);
  max = Math.ceil(max);

  for (let i = 0; i < arr.length; i++) {
    let n = arr[i];
    if (n < min) {
      console.log(n + ' < ' + min + '.')
      ok = false;
      break;
    }

    if (n > max) {
      console.log(n + ' > ' + max + '.')
      ok = false;
      break;
    }

		let count = 0;
    for (let j = 0; j < arr.length; j++) {
      if (n == arr[j]) {
				count++;
      }
    }

		if (count > 1) {
			console.log(n + ' repeats ' + count + ' times.')
			ok = false;
		}
  }

  return ok;
}

function runSingleTest(min: number, max: number, count?: number) {

  let hasCount: boolean;

	let min2 = Math.floor(min);
	let max2 = Math.ceil(max);
	let range = max2 - min2 + 1;

  if (typeof count === "undefined") {
    count = range;

    hasCount = false;
  } else {
    hasCount = true;
  }

  let arr = UniqueRand.getRandArr(min, max, count);
  let ok = isUnique(arr, min, max);

	if (count < 0) {
		ok = ok && (arr.length === 0)
	} else if (count > range) {
		ok = ok && (arr.length === range)
	} else {
		ok = ok && (arr.length === count)
	}

  if (hasCount) {
    console.log(`runSingleTest(${min}, ${max}, ${count}) = [${arr}] ... ${ok ? "Passed" : "Failed"}` );
  } else {
    console.log(`runSingleTest(${min}, ${max}) = [${arr}] ... ${ok ? "Passed" : "Failed"}`);
  }

  return ok;
}

test("all", () => {
  expect(runSingleTest(1, 1)).toBe(true);
  expect(runSingleTest(1, 2)).toBe(true);
  expect(runSingleTest(1, 3)).toBe(true);
  expect(runSingleTest(0, 10)).toBe(true);
  expect(runSingleTest(0.2, 9.3)).toBe(true);
  expect(runSingleTest(-0.2, 9.3)).toBe(true);
  expect(runSingleTest(-20.2, -9.3)).toBe(true);
  expect(runSingleTest(-5, 8)).toBe(true);
  expect(runSingleTest(1, 15)).toBe(true);
  expect(runSingleTest(5, 95)).toBe(true);

  expect(runSingleTest(0, 10, 5)).toBe(true);
  expect(runSingleTest(0.2, 9.3, 5)).toBe(true);
  expect(runSingleTest(-0.2, 9.3, 5)).toBe(true);
  expect(runSingleTest(-20.2, -9.3, 5)).toBe(true);
  expect(runSingleTest(-5, 8, 5)).toBe(true);
  expect(runSingleTest(1, 15, 6)).toBe(true);
  expect(runSingleTest(5, 95, 8)).toBe(true);

	// test count is > max - min + 1
  expect(runSingleTest(1, 5, 6)).toBe(true);

	// test count is > max - min + 1, with negative min
  expect(runSingleTest(-2, 5, 88)).toBe(true);

	// test count is < 0 
  expect(runSingleTest(1, 5, -8)).toBe(true);

	// test count is < 0, with negative min
  expect(runSingleTest(-8, 5, -8)).toBe(true);

	// test max < min
  expect(runSingleTest(8, -1, -8)).toBe(true);
});



UniqueRand.js
'use strict';
exports.__esModule = true;
exports.UniqueRand = void 0;
var UniqueRand = /** @class */ (function () {
    function UniqueRand() {
    }
    /* - returns a random number out of the universe
     * - removes it from the universe
     */
    UniqueRand.prototype.getRand = function (universe) {
        var max = universe.length;
        var i = Math.floor(Math.random() * max);
        var number = universe[i];
        universe.splice(i, 1);
        return number;
    };
    /* get array of random numbers between 0 and max - 1
     */
    UniqueRand.prototype.getUniqueRand = function (arr, universe, max) {
        if (0 == arr.length) {
            arr.push(this.getRand(universe));
            //console.log('first item ' + arr[0]);
        }
        else {
            var rand = this.getRand(universe);
            arr.push(rand);
        }
    };
    UniqueRand.prototype.getRandArr = function (min, max, count) {
        var arr = [];
        if (min > max) {
        }
        else {
            min = Math.floor(min);
            max = Math.ceil(max);
            var universe = [];
            var range = max - min + 1;
            if (typeof count === "undefined") {
                count = range;
            }
            for (var i = 0; i < range; i++) {
                universe.push(i);
            }
            for (var i = 0; i < count; i++) {
                this.getUniqueRand(arr, universe, range);
            }
            for (var i = 0; i < count; i++) {
                arr[i] += min;
            }
        }
        return arr;
    };
    return UniqueRand;
}());
exports.UniqueRand = UniqueRand;


uniqueRand.js
'use strict';

module.exports = {

  /* - returns a random number out of the universe
   * - removes it from the universe
   */
  getRand: function(universe) {
    let max = universe.length;
    let i = Math.floor(Math.random() * max)
    let number = universe[i];
    universe.splice(i, 1);
    return number;
  },

  /* get array of random numbers between 0 and max - 1
   */
  getUniqueRand: function(arr, universe, max) {
    if (0 == arr.length) {
      arr.push(this.getRand(universe));
      //console.log('first item ' + arr[0]);

    } else {
      let rand = this.getRand(universe);
      arr.push(rand);
    }
  },

  getRandArr: function(min, max) {
    
    let arr = [];

    if (min > max) {

    } else {
      min = Math.floor(min);
      max = Math.ceil(max);

      let universe = [];

      let range = max - min + 1;

      for (let i = 0; i < range; i++) {
        universe.push(i); 
      }

      for (let i = 0; i < range; i++) {
        this.getUniqueRand(arr, universe, range);
      }

      for (let i = 0; i < range; i++) {
        arr[i] += min;
      }
    }

    return arr;
  }
}




src/UniqueRand.ts
'use strict';

export class UniqueRand {

  /* - returns a random number out of the universe
   * - removes it from the universe
   */
  static getRand(universe: number[]): number {
    let max = universe.length;
    let i = Math.floor(Math.random() * max)
    let number = universe[i];
    universe.splice(i, 1);
    return number;
  }

  /* get array of random numbers between 0 and max - 1
   */
  static getUniqueRand(arr: number[], universe: number[], max: number) {
      arr.push(this.getRand(universe));
  }

  static getRandArr(min: number, max: number, count?: number): number[] {
    
    let arr: number[] = [];

    if (min > max) {
			return arr;
    } 

		min = Math.floor(min);
		max = Math.ceil(max);

		let universe = [];

		let range = max - min + 1;

		if (typeof count === "undefined") {
			count = range;
		} else {

			// Limit count to the maximum possible
			let max_count = max - min + 1;
			if (count > max_count) {
				count = max_count;
			}
		}

		for (let i = 0; i < range; i++) {
			universe.push(i); 
		}

		for (let i = 0; i < count; i++) {
			this.getUniqueRand(arr, universe, range);
		}

		for (let i = 0; i < count; i++) {
			arr[i] += min;
		}

    return arr;
  }
}




uniqueRandTest.ts
const { UniqueRand } = require('./UniqueRand');

function isUnique(arr: number[], min: number, max: number) {

  let ok = true;

  min = Math.floor(min);
  max = Math.ceil(max);

  for (let i = 0; i < arr.length; i++) {
    let n = arr[i];
    if (n < min) {
      console.log(n + ' < ' + min + '.')
      ok = false;
      break;
    }

    if (n > max) {
      console.log(n + ' > ' + max + '.')
      ok = false;
      break;
    }

    for (let j = 0; j < i; j++) {
      if (n == arr[j]) {
        console.log(n + ' repeats.')
        ok = false;
        break;
      }
    }
  }
  console.log(ok ? '  Passed' : '  Failed');
}

function runSingleTest(min: number, max: number, count?: number) {

  if (typeof count === "undefined") {
    let min2 = Math.floor(min);
    let max2 = Math.ceil(max);
    let range = max2 - min2 + 1;
    count = range;
  }

  let uniqueRandObj = new UniqueRand();
  let arr = uniqueRandObj.getRandArr(min, max, count);
  console.log('final arr = ' + arr);
  isUnique(arr, min, max);
}

function testRand() {
  runSingleTest(1, 1);
  runSingleTest(1, 2);
  runSingleTest(1, 3);
  runSingleTest(0, 10);
  runSingleTest(0.2, 9.3);
  runSingleTest(-0.2, 9.3);
  runSingleTest(-20.2, -9.3);
  runSingleTest(-5, 8);
  runSingleTest(1, 15);
  runSingleTest(5, 95);

  runSingleTest(0, 10, 5);
  runSingleTest(0.2, 9.3, 5);
  runSingleTest(-0.2, 9.3, 5);
  runSingleTest(-20.2, -9.3, 5);
  runSingleTest(-5, 8, 5);
  runSingleTest(1, 15, 6);
  runSingleTest(5, 95, 8);
}

testRand();


uniqueRandStandalone.js
'use strict';


/* - returns a random number out of the universe
 * - removes it from the universe
 */
function getRand(universe) {
  let max = universe.length;
  let i = Math.floor(Math.random() * max)
  let number = universe[i];
  universe.splice(i, 1);
  return number;
}

/* get array of random numbers between 0 and max - 1
 */
function getUniqueRand(arr, universe, max) {
  if (0 == arr.length) {
    arr.push(getRand(universe));
    //console.log('first item ' + arr[0]);

  } else {
    let rand = getRand(universe);
    arr.push(rand);
  }
}

function getRandArr(min, max) {
  
  let arr = [];

  if (min > max) {

  } else {
    min = Math.floor(min);
    max = Math.ceil(max);

    let universe = [];

    let range = max - min + 1;

    for (let i = 0; i < range; i++) {
      universe.push(i); 
    }

    for (let i = 0; i < range; i++) {
      getUniqueRand(arr, universe, range);
    }

    for (let i = 0; i < range; i++) {
      arr[i] += min;
    }
  }

  return arr;
}



uniqueRandTest.js
var UniqueRand = require('./UniqueRand').UniqueRand;
function isUnique(arr, min, max) {
    var ok = true;
    min = Math.floor(min);
    max = Math.ceil(max);
    for (var i = 0; i < arr.length; i++) {
        var n = arr[i];
        if (n < min) {
            console.log(n + ' < ' + min + '.');
            ok = false;
            break;
        }
        if (n > max) {
            console.log(n + ' > ' + max + '.');
            ok = false;
            break;
        }
        for (var j = 0; j < i; j++) {
            if (n == arr[j]) {
                console.log(n + ' repeats.');
                ok = false;
                break;
            }
        }
    }
    console.log(ok ? '  Passed' : '  Failed');
}
function runSingleTest(min, max, count) {
    if (typeof count === "undefined") {
        var min2 = Math.floor(min);
        var max2 = Math.ceil(max);
        var range = max2 - min2 + 1;
        count = range;
    }
    var uniqueRandObj = new UniqueRand();
    var arr = uniqueRandObj.getRandArr(min, max, count);
    console.log('final arr = ' + arr);
    isUnique(arr, min, max);
}
function testRand() {
    runSingleTest(1, 1);
    runSingleTest(1, 2);
    runSingleTest(1, 3);
    runSingleTest(0, 10);
    runSingleTest(0.2, 9.3);
    runSingleTest(-0.2, 9.3);
    runSingleTest(-20.2, -9.3);
    runSingleTest(-5, 8);
    runSingleTest(1, 15);
    runSingleTest(5, 95);
    runSingleTest(0, 10, 5);
    runSingleTest(0.2, 9.3, 5);
    runSingleTest(-0.2, 9.3, 5);
    runSingleTest(-20.2, -9.3, 5);
    runSingleTest(-5, 8, 5);
    runSingleTest(1, 15, 6);
    runSingleTest(5, 95, 8);
}
testRand();
